src/Razor/src/Microsoft.AspNetCore.Razor.Language/DefaultRazorTagHelperBinderPhase.cs

```diff
@@ -188,5 +255,155 @@ private bool AssemblyContainsTagHelpers(string assemblyName, IReadOnlyList<TagHe
+                    {
+                        // Make sure this node exists in the file we're parsing and not in its imports.
+                        if (_filePath.Equals(_source.FilePath, StringComparison.Ordinal))
+                        {
+                            tagHelperPrefix.Diagnostics.Add(
+                                ComponentDiagnosticFactory.Create_UnsupportedTagHelperDirective(node.GetSourceSpan(_source)));
+                        }
+                    }
+                    else if (context.ChunkGenerator is AddImportChunkGenerator usingStatement && !usingStatement.IsStatic)
+                    {
+                        // Get the namespace from the using statement. Split it at '=' in case there is an alias.
+                        var @namespace = usingStatement.ParsedNamespace.Split('=').LastOrDefault();
+                        for (var i = 0; i < _tagHelpers.Count; i++)
+                        {
+                            var tagHelper = _tagHelpers[i];
+                            if (tagHelper.IsComponentFullyQualifiedNameMatch())
+                            {
+                                // We've already added these to our list of matches.
+                                continue;
+                            }
+
+                            var typeName = tagHelper.GetTypeName();
+                            if (typeName != null && IsTypeInNamespace(typeName, @namespace))
+                            {
+                                // If the type is at the top-level or if the type's namespace matches the using's namespace, add it.
+                                Matches.Add(tagHelper);
+                            }
+                        }
+                    }
+                }
+            }
+
+            private bool IsTypeInNamespace(string typeName, string @namespace)
+            {
+                var namespaceLength = typeName.LastIndexOf('.');
+                if (namespaceLength == -1)
+                {
+                    // Either the typeName is not the full type name or this type is at the top level.
+                    return true;
+                }
+
+                var typeNamespace = typeName.Substring(0, namespaceLength);
+                if (typeNamespace.Equals(@namespace, StringComparison.Ordinal))
+                {
+                    return true;
+                }
+
+                return false;
+            }
+
+            private bool IsTypeInScope(string typeName, string currentNamespace)
+            {
+                var namespaceLength = typeName.LastIndexOf('.');
+                if (namespaceLength == -1)
+                {
+                    // Either the typeName is not the full type name or this type is at the top level.
+                    return true;
+                }
+
+                var typeNamespace = typeName.Substring(0, namespaceLength);
```

Comment:
I should probably add a comment here. What this is doing is checking if the given type is already in scope given the namespace of the current document.
E.g,
If the namespace of the document is `MyComponents.Components.Shared`,
then the types `MyComponents.FooComponent`, `MyComponents.Components.BarComponent`, `MyComponents.Components.Shared.BazComponent` are all in scope.
Whereas `MyComponents.SomethingElse.OtherComponent` is not in scope.

We need to do this so that we don't have to generate dynamic usings for each file.